runtime/runtime/src/bandwidth_scheduler/scheduler.rs

```diff
@@ -0,0 +1,765 @@
+//! shouldn't be too painful.
+
+use std::collections::{BTreeMap, VecDeque};
+use std::rc::Rc;
+
+use near_primitives::bandwidth_scheduler::{
+    Bandwidth, BandwidthRequest, BandwidthRequestValues, BandwidthRequests,
+    BandwidthSchedulerParams, BandwidthSchedulerState, BlockBandwidthRequests, LinkAllowance,
+};
+use near_primitives::types::ShardId;
+use rand::seq::SliceRandom;
+use rand::SeedableRng;
+use rand_chacha::ChaCha20Rng;
+
+use super::shard_mapping::{
+    SchedulerShardIndex, SchedulerShardIndexMap, SchedulerShardLinkMap, SchedulerShardMapping,
+    ShardLink,
+};
+
+/// How many bytes of outgoing receipts can be sent from one shard to another at the current height.
+/// Produced by the bandwidth scheduler.
+pub struct GrantedBandwidth {
+    granted: BTreeMap<(ShardId, ShardId), Bandwidth>,
+}
+
+impl GrantedBandwidth {
+    pub fn get_granted_bandwidth(&self, sender: ShardId, receiver: ShardId) -> Bandwidth {
+        self.granted.get(&(sender, receiver)).copied().unwrap_or(0)
+    }
+}
+
+pub struct BandwidthScheduler {
+    /// Mapping from shard ids to shard indexes.
+    mapping: SchedulerShardMapping,
+    /// List of shards that want to send receipts.
+    senders: Rc<[SchedulerShardIndex]>,
+    /// List of shards that the senders can send to.
+    receivers: Rc<[SchedulerShardIndex]>,
+    /// Configuration parameters for the algorithm.
+    params: BandwidthSchedulerParams,
+    /// ShardStatus for each shard.
+    shards_status: SchedulerShardIndexMap<ShardStatus>,
+    /// Each shard can send and receive at most `max_shard_bandwidth` bytes of receipts.
+    /// This is tracked in the `sender_budget` and `receiver_budget` fields, which keep
+    /// track of how much more a shard can send or receive before hitting the limit.
+    /// The sender and receiver budgets are decreased when bandwidth is granted on some link.
+    /// It's not possible to grant more bandwidth than the sender or receiver budget.
+    sender_budget: SchedulerShardIndexMap<Bandwidth>,
+    receiver_budget: SchedulerShardIndexMap<Bandwidth>,
+    /// Allowance for every link
+    /// Bandwidth scheduler uses `allowance` to ensure fairness.
+    /// Every link receives a fair amount of allowance on every height.
+    /// When bandwidth is granted on a link, the link's allowance is decreased
+    /// by the granted amount. Requests on links with higher allowance have priority
+    /// over requests on links with lower allowance.
+    /// Links that send more than their fair share are deprioritized, which keeps things fair.
+    link_allowances: SchedulerShardLinkMap<Bandwidth>,
+    /// How much bandwidth was granted on each link.
+    granted_bandwidth: SchedulerShardLinkMap<Bandwidth>,
+    /// Rng used to resolve ties when multiple requests have the same allowance.
+    rng: ChaCha20Rng,
+}
+
+impl BandwidthScheduler {
+    pub fn run(
+        sender_shards: &[ShardId],
```

Comment:
Ideally it would be
```rust
sender_layout: ShardLayout,
receiver_layout: ShardLayout
```
and we'd take the senders and receivers from the respective layouts.

For now this is out of scope, but we could do that in the future if it turns out that we really need 100% correct grants even on resharding boundaries. IMO it probably will stay as is for the foreseeable future :sweat_smile: 

I added a section about resharding in the module-level comment. Reading it again now, I admit that it isn't written as clearly as it could be, I'll try to rewrite it.